src/cmd_parser.cpp

```diff
@@ -110,28 +111,49 @@ bool CmdParser::parseNextCommand()
       d_state.pushScope();
       std::string name = d_eparser.parseSymbol();
       std::vector<Expr> vs =
-          d_eparser.parseAndBindSortedVarList();
-      // parse premises
-      std::string keyword = d_eparser.parseKeyword();
-      if (keyword!="premises")
-      {
-        d_lex.parseError("Expected premises in declare-rule");
-      }
-      std::vector<Expr> premises = d_eparser.parseExprList();
-      // parse args
-      keyword = d_eparser.parseKeyword();
-      if (keyword!="args")
+          d_eparser.parseAndBindSortedVarList(true);
+
+      std::vector<Expr> premises = std::vector<Expr>();
+      std::vector<Expr> args = std::vector<Expr>();
+      Expr conc = nullptr;
+
+      std::string keyword;
+      while (true)
       {
-        d_lex.parseError("Expected args in declare-rule");
+        Token tok = d_lex.peekToken();
+        if (tok == Token::RPAREN)
+        {
+          break;
+        }
+        std::string keyword = d_eparser.parseKeyword();
+        if (keyword == "premises")
+        {
+          premises = d_eparser.parseExprList();        
+          continue;
+        }
+        if (keyword == "args")
+        {
+          args = d_eparser.parseExprList();
+          continue;
+        }
+        if (keyword == "conclusion")
+        {
+          conc = d_eparser.parseExpr();
+          continue;
+        }
+        if (keyword == "require")
+        {
+          d_eparser.eatSexpr();
```

Comment:
I think we should parse an expression here and type check it.

I will add support for a command `program`.